core/src/main/java/org/fao/geonet/kernel/security/openidconnect/OidcUser2GeonetworkUser.java

```diff
@@ -0,0 +1,213 @@
+import org.fao.geonet.repository.UserGroupRepository;
+import org.fao.geonet.repository.UserRepository;
+import org.fao.geonet.repository.specification.UserGroupSpecs;
+import org.fao.geonet.utils.Log;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.oauth2.core.oidc.OidcIdToken;
+import org.springframework.util.StringUtils;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Class for handling Oidc User and the Geonetwork User.
+ * <p>
+ * Most of the work, here, is to save the user's profile-groups.  See SimpleOidcUser for other details.
+ */
+public class OidcUser2GeonetworkUser {
+
+    @Autowired
+    OIDCConfiguration oidcConfiguration;
+    @Autowired
+    OIDCRoleProcessor oidcRoleProcessor;
+    @Autowired
+    SimpleOidcUserFactory simpleOidcUserFactory;
+    @Autowired
+    private UserRepository userRepository;
+    @Autowired
+    private GroupRepository groupRepository;
+    @Autowired
+    private LanguageRepository langRepository;
+    @Autowired
+    private UserGroupRepository userGroupRepository;
+    @Autowired
+    private GeonetworkAuthenticationProvider geonetworkAuthenticationProvider;
+
+
+    public UserDetails getUserDetails(Map attributes, boolean withDbUpdate) {
+        SimpleOidcUser simpleUser = simpleOidcUserFactory.create(attributes);
+        if (!StringUtils.hasText(simpleUser.getUsername()))
+            return null;
+
+        User user;
+        boolean newUserFlag = false;
+        try {
+            user = (User) geonetworkAuthenticationProvider.loadUserByUsername(simpleUser.getUsername());
+        } catch (UsernameNotFoundException e) {
+            user = new User();
+            user.setUsername(simpleUser.getUsername());
+            newUserFlag = true;
+            Log.debug(Geonet.SECURITY, "Adding a new user: " + user);
+        }
+
+        simpleUser.updateUser(user); // copy attributes from the IDToken to the GN user
+
+        Map<Profile, List<String>> profileGroups = oidcRoleProcessor.getProfileGroups(attributes);
+        user.setProfile(oidcRoleProcessor.getProfile(attributes));
+
+
+        //Apply changes to database is required.
+        if (withDbUpdate) {
+            if (newUserFlag || oidcConfiguration.isUserProfileUpdateEnabled()) {
+                userRepository.save(user);
+            }
+            if (newUserFlag || oidcConfiguration.isUserGroupUpdateEnabled()) {
+                updateGroups(profileGroups, user);
+            }
+        }
+        return user;
+    }
+
+    /**
+     * given an oidc ID token, get the spring user details for the user.
+     * If this is a new user, it will be saved to the GN database (as a GN user).
+     * <p>
+     * If withDbUpdate is true, and  isUserProfileUpdateEnabled (oidc configuration),
```

Comment:
In keycloak, there was only one flag which was similar to "withDbUpdate"  and if "withDbUpdate" was set then isUserProfileUpdateEnabled would automatically be set to "false" because why would we want to allow users to edit their profile if it will just be overwritten each time they login?

I just did a test and I see that the isUserProfileUpdateEnabled defaults to true and allows me to update my profile.  But soon as I login again all my profile changes are rest to the values from keycloak.  This seems like a bad user experience.  I would recommend updating the logic so that if the data is coming from oidc then the user profile should not be editable.

Reply:
how do you turn off editing of the user profile?

Reply:
The following code from  class OIDCConfiguration is used to disable editing of the user profile.  If you compare this to keycloak then your function should `return (userProfileUpdateEnabled && !withDbUpdate)` which basically only allows the update to occur if we are not updating user profile in the database (and your new userProfileUpdateEnabled flag = true)
   ```
 @Override
    public boolean isUserProfileUpdateEnabled() {
        return userProfileUpdateEnabled;
    }
```

Reply:
Note that I'm not really sure I understand the purpose of the new variable "userProfileUpdateEnabled"

| userProfileUpdateEnabled|      withDbUpdate|  comment|
|----------|:-------------:|------:|
| true |  true | does not make sense because why would we want the user to update the profile when db update will simply overwrite the changes on next restart |
| true |  false| Make sense to update profile since dp update will not be changing anything |
| false |  false | Does not seem to make sense  because if the users notices an error in the user profile they cannot change it and it will never get fixed. Unless this is what we want to achieve? |
| false |  true | All changes are coming from openid so this makes sense |

Notice that for the ones that make sense - userProfileUpdateEnabled is the opposite of withDbUpdate.
So I don't quite understand why you would not want to simply return

`return !withDbUpdate;`

Which is what was done in keycloak

https://github.com/geonetwork/core-geonetwork/blob/de6702bda84e1b56c413cb6833a9e5dd0c8fb92a/core/src/main/java/org/fao/geonet/kernel/security/keycloak/KeycloakConfiguration.java#L143-L154

Also noticed that keycloak as "isUserGroupUpdateEnabled" which is based on "updateGroup" .  updateGroup is a flag that is used to indicate if the geonetwork groups should be updated based on keycloak settings on each login.  This is similar to withDbUpdate except for group settings.   So if we only wanted to get the group settings on keycloak first login and then allow someone else to change it then updateGroup would be set to false. To be honest, I don't understand when this would ever occur?  (I sometimes think that isUserGroupUpdateEnabled should just be hardcoded to return "false" since I would expect that the groups will always come from keycloak/oidc so why allow administrators to change the group assignments (I'm not sure I know of a use case)?)

Note that the updateGroup and updateProfile are not new variables that were created for keycloak.   The logic was copied from Shibboleth.
https://github.com/geonetwork/core-geonetwork/blob/de6702bda84e1b56c413cb6833a9e5dd0c8fb92a/core/src/main/java/org/fao/geonet/kernel/security/shibboleth/ShibbolethUserUtils.java#L155-L168

So I would expect that your code should have used similar naming to avoid confusion.  Or alternatively, you could rename the variables so that they all align.

Hopefully this helps explain the history of these variables....

Reply:
Hmmm - complicated.  Thanks for the summary - that helps a lot. I think I had mis-understood them.

It seems like there should only be 2 option (the ones that you said made sense).

I've changed to be consistent with keycloak.